packages/flutter/lib/src/material/pickers/calendar_date_picker.dart

```diff
@@ -0,0 +1,1050 @@
+  final DateTime initialDate;
+
+  /// The earliest allowable [DateTime] that the user can select.
+  final DateTime firstDate;
+
+  /// The latest allowable [DateTime] that the user can select.
+  final DateTime lastDate;
+
+  /// Called when the user selects a date in the picker.
+  final ValueChanged<DateTime> onDateChanged;
+
+  /// Called when the user navigates to a new month/year in the picker.
+  final ValueChanged<DateTime> onDisplayedMonthChanged;
+
+  /// The initial display of the calendar picker.
+  final DatePickerMode initialCalendarMode;
+
+  /// Function to provide full control over which dates in the calendar can be selected.
+  final SelectableDayPredicate selectableDayPredicate;
+
+  @override
+  _CalendarDatePickerState createState() => _CalendarDatePickerState();
+}
+
+class _CalendarDatePickerState extends State<CalendarDatePicker> {
+  bool _announcedInitialDate = false;
+  DatePickerMode _mode;
+  DateTime _currentDisplayedMonthDate;
+  DateTime _nextMonthDate;
+  DateTime _previousMonthDate;
+  DateTime _selectedDate;
+  PageController _pageController;
+  final GlobalKey _pickerKey = GlobalKey();
+  MaterialLocalizations _localizations;
+  TextDirection _textDirection;
+
+  @override
+  void initState() {
+    super.initState();
+    _mode = widget.initialCalendarMode;
+    _currentDisplayedMonthDate = DateTime(widget.initialDate.year, widget.initialDate.month);
+    _previousMonthDate = utils.addMonthsToMonthDate(_currentDisplayedMonthDate, -1);
+    _nextMonthDate = utils.addMonthsToMonthDate(_currentDisplayedMonthDate, 1);
+    _selectedDate = widget.initialDate;
+    _pageController = PageController(initialPage: utils.monthDelta(widget.firstDate, _currentDisplayedMonthDate));
+  }
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    _localizations = MaterialLocalizations.of(context);
+    _textDirection = Directionality.of(context);
+    if (!_announcedInitialDate) {
+      _announcedInitialDate = true;
+      SemanticsService.announce(
+        _localizations.formatFullDate(_selectedDate),
+        _textDirection,
+      );
+    }
+  }
+
+  void _vibrate() {
+    switch (Theme.of(context).platform) {
+      case TargetPlatform.android:
+      case TargetPlatform.fuchsia:
+      case TargetPlatform.linux:
+      case TargetPlatform.windows:
+        HapticFeedback.vibrate();
+        break;
+      case TargetPlatform.iOS:
+      case TargetPlatform.macOS:
+        break;
+    }
+  }
+
+  void _setCurrentMonth(DateTime date) {
+    setState(() {
+      if (_currentDisplayedMonthDate.year != date.year || _currentDisplayedMonthDate.month != date.month) {
+        _currentDisplayedMonthDate = DateTime(date.year, date.month);
+        _previousMonthDate = utils.addMonthsToMonthDate(_currentDisplayedMonthDate, -1);
+        _nextMonthDate = utils.addMonthsToMonthDate(_currentDisplayedMonthDate, 1);
+        _pageController?.dispose();
```

Comment:
This seems a bit risky. Presumably there's no way for the now-disposed page controller to be accessed, even though it's part of the current widget tree.  Similarly, I assume that there's no way for _setCurrentMonth() to be called more than once in between builds - so we will not unnecessarily dispose/create page controllers?

Reply:
Indeed. I just refactored the page controller out of the CalendarDatePicker and into the _MonthPicker so this will no longer be an issue.